vendor/zkllvm-metacraft-circuits/src/circuits_imp/weigh_justification_and_finalization.h

```diff
@@ -0,0 +1,327 @@
+                                         const CheckpointVariable& previous_justified_checkpoint,
+                                         const CheckpointVariable& current_justified_checkpoint,
+                                         const Epoch& current_epoch,
+                                         const CheckpointVariable& finalized_checkpoint) {
+    const uint64_t one = 1;
+    const uint64_t two = 2;
+    const uint64_t three = 3;
+
+    const auto bits_set_1_through_4_pred = justification_bits.test_range(1, 4);
+    const auto bits_set_1_through_3_pred = justification_bits.test_range(1, 3);
+    const auto bits_set_0_through_3_pred = justification_bits.test_range(0, 3);
+    const auto bits_set_0_through_2_pred = justification_bits.test_range(0, 2);
+
+    const auto previous_justified_checkpoint_epoch_plus_three = previous_justified_checkpoint.epoch + three;
+    const auto previous_justified_checkpoint_epoch_plus_two = previous_justified_checkpoint.epoch + two;
+    const auto current_justified_checkpoint_epoch_plus_two = current_justified_checkpoint.epoch + two;
+    const auto current_justified_checkpoint_epoch_plus_one = current_justified_checkpoint.epoch + one;
+
+    const auto second_using_fourth_as_source_pred = previous_justified_checkpoint_epoch_plus_three == current_epoch;
+
+    const auto second_using_third_as_source_pred = previous_justified_checkpoint_epoch_plus_two == current_epoch;
+
+    const auto first_using_third_as_source_pred = current_justified_checkpoint_epoch_plus_two == current_epoch;
+
+    const auto first_using_second_as_source_pred = current_justified_checkpoint_epoch_plus_one == current_epoch;
+
+    const auto should_finalize_previous_justified_checkpoint_1_pred =
+        bits_set_1_through_4_pred && second_using_fourth_as_source_pred;
+
+    const auto should_finalize_previous_justified_checkpoint_2_pred =
+        bits_set_1_through_3_pred && second_using_third_as_source_pred;
+
+    const auto should_finalize_previous_justified_checkpoint_pred =
+        should_finalize_previous_justified_checkpoint_1_pred || should_finalize_previous_justified_checkpoint_2_pred;
+
+    const auto should_finalize_current_justified_checkpoint_1_pred =
+        bits_set_0_through_3_pred && first_using_third_as_source_pred;
+
+    const auto should_finalize_current_justified_checkpoint_2_pred =
+        bits_set_0_through_2_pred && first_using_second_as_source_pred;
+
+    const auto should_finalize_current_justified_checkpoint_pred =
+        should_finalize_current_justified_checkpoint_1_pred || should_finalize_current_justified_checkpoint_2_pred;
+
+    auto new_finalized_checkpoint =
+        should_finalize_previous_justified_checkpoint_pred ? previous_justified_checkpoint : finalized_checkpoint;
+
+    new_finalized_checkpoint =
+        should_finalize_current_justified_checkpoint_pred ? current_justified_checkpoint : new_finalized_checkpoint;
+
+    return new_finalized_checkpoint;
+}
```

Comment:
Make this look like it does in the consensus spec, no need for these predicates and ternaries